linc_convert/modalities/lsm/transfer.py

```diff
@@ -0,0 +1,106 @@
+import dandi.upload
+
+# internals
+from linc_convert.modalities.lsm.cli import lsm
+
+transfer = cyclopts.App(name="transfer", help_format="markdown")
+lsm.command(transfer)
+
+@transfer.default
+def dandi_transfer(input_dir, dandiset_url, dandi_instance, subject, output_dir='.',  max_size_gb=2.00, upload=False):
+    """
+    Upload .dat files to DANDI in batched, compressed tar archives.
+    
+    Parameters
+    ----------
+    input_dir : str
+        Directory containing .dat files to upload
+    dandiset_url : str
+        URL for the dandiset to upload (e.g., https://lincbrain.org/dandiset/000010)
+    dandi_instance : str
+        DANDI server (e.g. linc, dandi)
+    output_dir : str, optional
+        Directory to save the Dandiset directory (default: '.')
+    max_size_gb : float, optional
+        Maximum size for each archive in GB (default: 2)
+    upload : bool, optional
+        Upload data to DANDI (default: True)
+    """
+
+    max_size_bytes = int(max_size_gb * 1024 * 1024 * 1024)
+
+    dandi.download.download(dandiset_url, output_dir=output_dir)
+
+    dandiset_id = dandiset_url.split('/')[-1]
+    dandiset_directory = f'{output_dir}/{dandiset_id}'
+
+    if not os.path.exists(f'{dandiset_directory}/dataset_description.json'):
+        with open(f'{dandiset_directory}/dataset_description.json', 'w') as f:
+            f.write('{}')
+
+    archive_directory = f'{dandiset_directory}/sourcedata/sub-{subject}'
+    os.makedirs(archive_directory, exist_ok=True)
+
+    dat_files = list(Path(input_dir).glob("*.dat"))
+    dat_files_size = len(dat_files)
+    if dat_files_size:
+        print(f"Found {dat_files_size} .dat files in '{input_dir}'.")
+    else:
+        print(f"No .dat files found in '{input_dir}'.")
+        return
+
+    batch = 0
+    file_number = 0
+
+    while file_number < dat_files_size:
+
+        print(f"\n[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] Creating archive batch {batch}")
+
+        archive_path = os.path.join(archive_directory, f"sub-{subject}_desc-batch{batch}.tar")
+        archive = tarfile.open(archive_path, "w")
+        
+        batch_size = 0
+        batch_files = 0
+
+        while batch_size < max_size_bytes and file_number < dat_files_size:
+
+            file_path = dat_files[file_number]
+            file_size = os.path.getsize(file_path)
+            
+            print(f"Adding '{file_path.name}' ({file_size/1024**2:.2f}MB, {file_number}) to archive.")
+            archive.add(file_path, arcname=file_path.name)
+
+            batch_size += file_size
+            batch_files += 1
+            file_number += 1
+
+        archive.close()
+
+        print(f"Archive created with {batch_files} files and {batch_size / 1024**2:.2f}MB size.")
+
+        if upload:
+            print(f"Uploading {archive_path}.")
+            dandi.upload.upload([dandiset_directory],
+                                dandi_instance=dandi_instance,
+                                )
```

Comment:
Is this robust enough? Would it be useful to wrap this into a loop with try/except and allow for multiple tries?

Reply:
Also, maybe we should have an option to restart a transfer where it was left off (say that the person stops it with ctrl+c). Not sure how to implement this.

Reply:
> Is this robust enough? Would it be useful to wrap this into a loop with try/except and allow for multiple tries?

Good idea.  I have added a loop to continually attempt the upload until successful.

Reply:
> Also, maybe we should have an option to restart a transfer where it was left off (say that the person stops it with ctrl+c). Not sure how to implement this.

I could create a manifest file that is stored locally to track the files that have been added to a tar and the tars that have been uploaded.